File: src/go/problems/min_reorder_test.go
// https://leetcode.com/problems/reorder-routes-to-make-all-paths-lead-to-the-city-zero

package problems

import (
	"testing"
	"github.com/dobermanch/leetcode/core"
)

type MinReorder struct{}

func TestMinReorder(t *testing.T) {
	gen := core.TestSuite[MinReorder]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param(10).Param(13).Result(23)
	}).Add(func(tc *core.TestCase) {
		tc.Param(-10).Param(12).Result(2)
	}).Run(t)
}

func (MinReorder) Solution(n int, connections [][]int) int {	
	neighbors := map[int][][]int{}
	for i := 0; i < n; i++ {
		neighbors[i] = [][]int{}
	}

	for i := 0; i < len(connections); i++ {
		neighbors[connections[i][0]] = append(neighbors[connections[i][0]], []int{connections[i][1], 1})
		neighbors[connections[i][1]] = append(neighbors[connections[i][1]], []int{connections[i][0], 0})
	}

	visited := make([]bool, n)
	
	var dfs func(city int) int
	dfs = func(current int) int {
		result := 0
		visited[current] = true

		for i := 0; i < len(neighbors[current]); i++ {
			neighbor := neighbors[current][i]
			if visited[neighbor[0]] {
				continue
			}

			if neighbor[1] == 1 {
				result++
			}

			result += dfs(neighbor[0])
		}
		return result
	}

	return dfs(0)
}

File: src/python/problems/min_reorder_test.py
#https://leetcode.com/problems/reorder-routes-to-make-all-paths-lead-to-the-city-zero

from core.problem_base import *

class MinReorder(ProblemBase):
    def Solution(self, n: int, connections: list[list[int]]) -> int:
        def Dfs(current):
            result = 0
            visited.add(current)
            for city, change in neighbors[current]:
                if city in visited:
                    continue

                if change:
                    result += 1
                
                result += Dfs(city)
            return result

        visited = set()
        neighbors = { city:[] for city in range(n) }
        for city1, city2 in connections:
            neighbors[city1].append((city2, True))
            neighbors[city2].append((city1, False))
        
        return Dfs(0)
        

if __name__ == '__main__':
    TestGen(MinReorder) \
        .Add(lambda tc: tc.Param(6).Param([[0,1],[1,3],[2,3],[4,0],[4,5]]).Result(3)) \
        .Add(lambda tc: tc.Param(5).Param([[1,0],[1,2],[3,2],[3,4]]).Result(2)) \
        .Add(lambda tc: tc.Param(3).Param([[1,0],[2,0]]).Result(0)) \
        .Run()
